CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

IntegralProgrammer opened the pull request:
This PR adds the `Utilities/Administration/github_publish_os_package_list.py` script which detects the host's OS as either _Debian_ or _Ubuntu_ and publishes the list of installed apt/dpkg packages to a GitHub repository.

### Setup / Testing Instructions

1. Create a new GitHub repository to store the apt/dpkg status files. In this example, we'll call it `CARDS-1892-test`. Initialize the repository with a _README_ file.
2. Create the following _empty_ files in that repository:
  - `hosts/cardsdemo.uhndata.io/vm/debian_version`
  - `hosts/cardsdemo.uhndata.io/vm/os-release`
  - `hosts/cardsdemo.uhndata.io/vm/lsb-release`
  - `hosts/cardsdemo.uhndata.io/vm/status`
3. Create a new GitHub App by going to https://github.com/settings/apps and click the _New GitHub App_ button.
4. Give your application some random name, set the _Homepage URL_ your GitHub page (eg. https://github.com/IntegralProgrammer), disable _Expire user authentication tokens_, deactivate the _webhook_, set the _Single file_ _Read and write_ permissions for the following files:
  - `hosts/cardsdemo.uhndata.io/vm/debian_version`
  - `hosts/cardsdemo.uhndata.io/vm/os-release`
  - `hosts/cardsdemo.uhndata.io/vm/lsb-release`
  - `hosts/cardsdemo.uhndata.io/vm/status`

![Screenshot_2022-09-08_13-06-06](https://user-images.githubusercontent.com/5349728/189183460-f6621101-d567-4358-a1be-c4436898558a.png)

Set _Where can this GitHub App be installed?_ to _Only on this account_ and click _Create GitHub App_.

5. Click the _Generate a private key_ button and save the private key `.pem` file.
6. Make note of the _App ID_
7. Install the App into the `CARDS-1892-test` repository.
![Screenshot_2022-08-19_13-26-51](https://user-images.githubusercontent.com/5349728/189184397-714fb41a-ba2a-4715-8f99-c9a21e3b2b27.png)

8. Make note of the _installation ID_ from the URL in the address bar `https://github.com/settings/installations/<INSTALLATION ID IS HERE>`.
9. `docker run --rm --name debian-test -it debian:11`
10. Within that container run `apt update && apt install python3 python3-jwt python3-distro python3-requests`
11. Outside of that container, use `docker cp` to copy the private key PEM file, the `github_publish_os_package_list.py` file, and the `os_file_management.py` file into the root of the `debian-test` container.
13. Inside the `debian-test` container run `python3 github_publish_os_package_list.py --private_key /PATH/TO/PRIVATE-KEY.PEM/FILE --app_id APP-ID-FROM-STEP-6 --installation_id INSTALLATION-ID-FROM-STEP-8 --repository GITHUB-USERNAME/CARDS-1892-test --deployment_hostname cardsdemo.uhndata.io`.
14. Check that the `debian_version`, `os-release`, and `status` files on GitHub have been updated.
15. Repeat steps 9 through 13 again but this time use an `ubuntu:18.04` image instead of `debian:11`
16. Check that the `lsb-release`, `os-release`, and `status` files on GitHub have been updated.

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Add the Utilities/Administration/github_publish_os_package_list.py script

sdumitriu commented on Utilities/Administration/github_publish_os_package_list.py:
```diff
@@ -0,0 +1,85 @@
+installation_token_resp = requests.post("https://api.github.com/app/installations/" + args.installation_id + "/access_tokens", headers=application_headers)
+installation_token = installation_token_resp.json()['token']
+
+if THIS_DISTRO == 'debian':
+	updateFile(args.deployment_hostname + "/vm/os-release", "/usr/lib/os-release", installation_token)
+	updateFile(args.deployment_hostname + "/vm/debian_version", "/etc/debian_version", installation_token)
+	updateFile(args.deployment_hostname + "/vm/status", "/var/lib/dpkg/status", installation_token)
+elif THIS_DISTRO == 'ubuntu':
+	updateFile(args.deployment_hostname + "/vm/os-release", "/usr/lib/os-release", installation_token)
+	updateFile(args.deployment_hostname + "/vm/lsb-release", "/etc/lsb-release", installation_token)
+	updateFile(args.deployment_hostname + "/vm/status", "/var/lib/dpkg/status", installation_token)
```
Would it make sense to do this instead:

```suggestion
if THIS_DISTRO == 'debian' || THIS_DISTRO == 'ubuntu':
	updateFile(args.deployment_hostname + "/vm/os-release", "/usr/lib/os-release", installation_token)
	updateFile(args.deployment_hostname + "/vm/os-version", "/etc/debian_version" if THIS_DISTRO == 'debian' else "/etc/lsb-release", installation_token)
	updateFile(args.deployment_hostname + "/vm/status", "/var/lib/dpkg/status", installation_token)
```

There's less duplicated code, and there's only one file on our end holding the version.

IntegralProgrammer replied:
I've addressed this code duplication issue in commit https://github.com/data-team-uhn/cards/pull/1158/commits/6a5114999ccaae25daec070455394c302d48cd09 by adding the `os_file_management.py` simple module. I don't want to use the generic `/vm/os-version` solution that you have proposed as, in doing so, we loose the information of what the original filename was and thus when we need to recreate the simple file structure in a test Docker image (CARDS-1896) we will not know what `/vm/os-version` should be renamed to - `/etc/debian_version` or `/etc/lsb-release`?

sdumitriu replied:
OK, sounds good.

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Address code review comment by de-duplicating calls to updateFile through the use of the OS_FILE_SET map defined in the os_file_management.py file

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Change the path on GitHub from DEPLOYMENT_HOSTNAME/vm/... to hosts/DEPLOYMENT_HOSTNAME/vm/...

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Refactor github_publish_os_package_list.py by moving the GitHub API interaction code into GitHubRepoHandler.py as future scripts to be created under CARDS-1701 will need to use this functionality

IntegralProgrammer committed:
CARDS-1892: Publish the Debian/Ubuntu APT package status files for the VM's OS, for each VM that we are using for a CARDS deployment

Move to Utilities/SecurityScanning

IntegralProgrammer commented:
Closing as the CARDS-1892 task is implemented in https://github.com/data-team-uhn/cards/pull/1179
